### ISSUE TITLE
invalid_client bad credentials when trying to use password grant_type on public client

### ISSUE THREAD

**Issue by davydotcom** [bluesliverx/grails-spring-security-oauth2-provider#98]
WARN  happyapps.GormUserDetailsService  - User not found: my-client-id

I have a client id and it appears its trying to authenticate the client id against my user database and not the Client model even though the config is 

Grails: 2.5.0
:spring-security-oauth2-provider:2.0-RC4
'org.grails.plugins:spring-security-core:2.0-RC3

```groovy
grails.plugin.springsecurity.oauthProvider.clientLookup.className = 'com.happyapps.Client'
grails.plugin.springsecurity.oauthProvider.authorizationCodeLookup.className = 'com.happyapps.AuthorizationCode'
grails.plugin.springsecurity.oauthProvider.accessTokenLookup.className = 'com.happyapps.AccessToken'
grails.plugin.springsecurity.oauthProvider.refreshTokenLookup.className = 'com.happyapps.RefreshToken'
grails.plugin.springsecurity.filterChain.chainMap = [
            '/oauth/token': 'JOINED_FILTERS,-oauth2ProviderFilter,-securityContextPersistenceFilter,-logoutFilter,-rememberMeAuthenticationFilter,-exceptionTranslationFilter',
            '/api/**': 'JOINED_FILTERS,-securityContextPersistenceFilter,-logoutFilter,-rememberMeAuthenticationFilter,-basicAuthenticationFilter,-exceptionTranslationFilter',
            '/**': 'JOINED_FILTERS,-statelessSecurityContextPersistenceFilter,-oauth2ProviderFilter,-clientCredentialsTokenEndpointFilter,-basicAuthenticationFilter,-oauth2ExceptionTranslationFilter'
   ]
```

I have set the client_secret to blank which has caused an encoded blank password to be set. I have also setup its authorized grant types to allow all grant types

**Comment by bobby-vandiver** [comment]
Have you added `clientCredentialsAuthenticationProvider` to your providers in `Config.groovy` per [the documentation](http://bluesliverx.github.io/grails-spring-security-oauth2-provider/guide/gettingStarted.html#addClientProvider)? You should have something like the following:

```groovy
grails.plugin.springsecurity.providerNames = [
        'clientCredentialsAuthenticationProvider',
        'daoAuthenticationProvider',
        'anonymousAuthenticationProvider',
        'rememberMeAuthenticationProvider'
]
```

**Comment by davydotcom** [comment]
Yes I have this

```groovy
grails.plugin.springsecurity.providerNames = [
        'clientCredentialsAuthenticationProvider',
        'daoAuthenticationProvider',
        'anonymousAuthenticationProvider',
        'rememberMeAuthenticationProvider'
]
```
--
David Estes

On June 21, 2015 at 3:49:14 PM, Bobby Vandiver (notifications@github.com) wrote:

Have you added clientCredentialsAuthenticationProvider to your providers in Config.groovy per the documentation? You should have something like the following:

grails.plugin.springsecurity.providerNames = [
        'clientCredentialsAuthenticationProvider',
        'daoAuthenticationProvider',
        'anonymousAuthenticationProvider',
        'rememberMeAuthenticationProvider'
]

**Comment by davydotcom** [comment]
Further Details:

I just attempted to update to the RC5 spring security grails plugin and a weird behavior change happened. Now it looks like the clientCredentials filter is processing however its processing for everything even standard login (non oauth) . Also when i try to do password grant type it appears it tries to process the `username` param as the clientId
```groovy
Message: No client with requested id: davydotcom
    Line | Method
->>   80 | attemptAuthentication in grails.plugin.springsecurity.web.authentication.RequestHolderAuthenticationFilter
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
|     53 | doFilter              in     ''
|     62 | doFilter . . . . . .  in grails.plugin.springsecurity.web.authentication.logout.MutableLogoutFilter
|     67 | doFilter              in spud.permalinks.PermalinkFilter
|     27 | doFilter . . . . . .  in spud.core.SpudMultiSiteFilter
|   1145 | runWorker             in java.util.concurrent.ThreadPoolExecutor
|    615 | run . . . . . . . . . in java.util.concurrent.ThreadPoolExecutor$Worker
^    744 | run                   in java.lang.Thread
Caused by NoSuchClientException: No client with requested id: davydotcom
->>   25 | loadClientByClientId  in grails.plugin.springsecurity.oauthprovider.GormClientDetailsService
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 
|     80 | attemptAuthentication in grails.plugin.springsecurity.web.authentication.RequestHolderAuthenticationFilter
|     53 | doFilter . . . . . .  in     ''
|     62 | doFilter              in grails.plugin.springsecurity.web.authentication.logout.MutableLogoutFilter
|     67 | doFilter . . . . . .  in spud.permalinks.PermalinkFilter
|     27 | doFilter              in spud.core.SpudMultiSiteFilter
|   1145 | runWorker . . . . . . in java.util.concurrent.ThreadPoolExecutor
|    615 | run                   in java.util.concurrent.ThreadPoolExecutor$Worker
^    744 | run . . . . . . . . . in java.lang.Thread
```

**Comment by davydotcom** [comment]
Its as if the params.username is being used for both filters as opposed to the client credentials using params.client_id

**Comment by davydotcom** [comment]
Ok got everything to work !! Solution is I backed down spring-security-core to 2.0-RC4 instead of RC5 somethings up with the integration between RC4 and RC5

**Comment by bobby-vandiver** [comment]
Correct me if I'm wrong, @bluesliverx, but we haven't released an RC5. At any rate, it sounds like you got this resolved, @davydotcom. Closing this issue.

**Comment by bluesliverx** [comment]
@bobby-vandiver, I believe the issue was with the core plugin's RC5.  I opened up a new ticket to track the issue.

**Comment by syntheticzero** [comment]
I just want to say --- I'm running into this same bug. core-RC5 breaks the OAuth provider plugin. This is a pretty serious bug, since OAuth is a major use case.

**Comment by syntheticzero** [comment]
ps I'm glad I ran across this ticket since I was banging my head against the wall trying to figure out what I was doing wrong --- downgrading back to RC4 "fixed" the problem, but obviously I hope this bug is fixed for RC6 and later.

**Comment by bobby-vandiver** [comment]
@syntheticzero Have you tried pulling in the `SNAPSHOT` version of the OAuth 2 provider plugin? It should be compatible with Spring Security Core RC5. 

I was hoping to get some of the outstanding issues resolved before we release RC5 of the plugin but maybe it's best to go ahead and do that to address these incompatibility issues. 

@bluesliverx What are your thoughts?

**Comment by bluesliverx** [comment]
Yup, I was just waiting on word from you to release it.  I can release RC5, you ok with that?

**Comment by bobby-vandiver** [comment]
It should be safe. I did a quick look over the open issues and most of them are for customization or other non critical things.

**Comment by bluesliverx** [comment]
I released RC5 yesterday.
